### ISSUE TITLE
Establish a semver major release schedule

### ISSUE THREAD

**Issue by cjihrig** [libuv/libuv#1597]
I'd like to propose that we start making semver major releases at least once a year, assuming there are semver major changes available. We could adopt a more frequent schedule like Node's, but I don't think we have the amount of churn (plus the V8 dependency) to necessitate that.

Currently, the `v2` label makes up ~15% of all of our open issues and PRs. As more of these v2 changes get queued up, it makes it harder to get all of these changes reviewed and merged, pushing back the v2 release. Alternatively, if we cut v2 without all of these breaking changes included, then the ones that are left behind for v3 could sit for years.

cc: @libuv/collaborators

**Comment by bnoordhuis** [comment]
I agree.  Stacking up lots of big changes means making people upgrade is going to be a protracted battle.

**Comment by santigimeno** [comment]
I think this is a good idea.

**Comment by cjihrig** [comment]
How do you feel about a December, January, or February timeframe? That would give us a few months to get things in order, but be sure not to align with a Node semver-major. I'd like to avoid "let's postpone the Node release for the next libuv major" if possible.

**Comment by bnoordhuis** [comment]
Sounds good to me.

**Comment by refack** [comment]
I'm assuming ABI compatibility is still a goal?

**Comment by bnoordhuis** [comment]
Not between v1 and v2, no.

**Comment by bzoz** [comment]
SGTM

**Comment by skypjack** [comment]
ABI compatibility not guaranteed, it makes sense. Breaking changes and changes to the API in general will be fully tracked and documented? I mean something like [this](https://abi-laboratory.pro/tracker/timeline/libuv/), far more handful than a raw changelog.  
I'm asking this for it could help to keep up-to-date external projects.

**Comment by bnoordhuis** [comment]
We'll call out the big changes in the release notes but every last little change?  Probably not, unless someone volunteers to put together an overview.

It could go on the wiki as a collaborative effort, that's what we do for Node.js.

**Comment by skypjack** [comment]
Do you think the ABI tracker above won't work? If it tracked also the wip branch it would be enough.  
I have no idea of how it works, just asking.  
My thought is that keeping the _projects that use libuv_ updated helps to fill the gap as soon as v2 is ready. That's all.

**Comment by santigimeno** [comment]
SGTM

**Comment by bnoordhuis** [comment]
@skypjack Yep, abi-tracker should catch such changes.  I was replying to your question whether all such changes will be documented by us (as in: in a manually written, copy-edited document.)

**Comment by skypjack** [comment]
@bnoordhuis I see. No, I didn't dare to ask that much!! My only concern is that abi-tracker plots the changes as a whole when the releases are _published_. To keep updated external projects during the development it would help a kind of step-by-step tracking mechanism. I mean, something like abi-tracker launched once a week or whatever isn't a burden for you all.  
Do you think it's possible?

**Comment by bnoordhuis** [comment]
Certainly, the software is on GitHub: https://github.com/lvc/abi-tracker

(That said, I'm not volunteering to set it up.)

**Comment by skypjack** [comment]
Got it. So, the answer is - _no, it won't happen, but feel free to do it for yourself_. Right?  
Fair enough. Thank you for the link.

**Comment by jebej** [comment]
It seems that someone is already maintaining a report here: https://abi-laboratory.pro/tracker/timeline/libuv/index.html

**Comment by skypjack** [comment]
@jebej 
Yep, that's the ABI tracker we mentioned in the previous comments. It tracks changes when a new version of `libuv` is released.
The question was about tracking changes on the v2.x development branch periodically instead, before an actual release is available. It would help keeping updated the projects that use `libuv`.

**Comment by jebej** [comment]
Ah I hadn't seen the link. Sorry about the noise...

**Comment by cjihrig** [comment]
@libuv/collaborators assuming that `master` will be synced with `v1.x` via https://github.com/libuv/libuv/pull/1753 in the next day or so, would anyone be opposed to setting Monday, March 12th as a target date for v2.0.0?

Once #1753 lands, it should just be a matter of landing some/most/all of the PRs labeled v2.

**Comment by davisjam** [comment]
I'd love to squeeze the new threadpool into v2. I have a paper deadline March 9 and will have two weeks off after that in which I think I could get the threadpool done if reviewers were willing to play along. This would require postponing the release until, say, the 23rd though.

**Comment by bnoordhuis** [comment]
I don't want to rush a big change right before a major release but you can claim stakes where necessary (e.g. struct sizes) so that it can land in a minor release.

> would anyone be opposed to setting Monday, March 12th as a target date for v2.0.0?

Well... the thought struck me that we'd have to maintain v1.x until Q2 2021 if node 10 were to ship with it.  If we can't get node upgraded before the release that makes v2.x rather less appealing  because of the double maintenance.

**Comment by cjihrig** [comment]
@jasnell are you the one cutting the Node 10 release? If so, when is the latest we could release libuv v2 and make it into 10? Please keep in mind that if libuv 2 can't make it into Node 10, it could potentially stall the release for a long time.

I'm willing to work more quickly to land as many of the v2 PRs as possible, but I'm not sure what fun might be waiting when it comes time to land it in Node.

**Comment by jasnell** [comment]
There is one thing that I'd still like to get in to libuv 2 before it's cut. I can work on it this week. 

In generally, end of March is the latest to get it in to node.js 10.0.0

**Comment by bnoordhuis** [comment]
Okay, so that's a ~2 week window.  Seems risky.  I don't think there's enough meat in master to really justify that.  Can I suggest we postpone and make it coincide with node 12?

Commits that are in master but not v1.x:
```
$ git log --format='%h %s' --no-merges origin/master ^origin/v1.x -- include/ src/
70d61b6c3 win: map ERROR_ACCESS_DENIED to UV_EACCES
d010030ad core: move all include files except uv.h to uv/
9834c4f75 src: return UV_EINVAL in uv_##name##_start()
78b8bcc3e win: adapt to new request initialization define
f65e004aa unix,win: remove v1.x version of uv_get_osfhandle
68dafe901 win: LEP-0005: remove uv_file
384a727a8 src: handle if_nametoindex errors in uv_ip6_addr
66319cf49 unix: fail with ENAMETOOLONG in `uv_pipe_*`
239ab6b01 win: map ERROR_INVALID_NAME to UV_EINVAL
912d42b72 unix,win: unify timers implementation
daf8a45d4 win: return IOCP HANDLE in uv_backend_fd
1f21b19e0 win: simplify global loops queue
d527ab90f core: switch to C99
ea13b6a99 win: remove no longer needed define
db83ffe3b win: fix compilation warning
250eaf439 Revert "win,tty: remove no longer needed workaround"
6947328d6 win: remove types already defined in MinGW-w64
56292929a win: simplify uv_rwlock_t type
b3c9902df win: simplify uv_cond_t type
058d94b2d win: remove unneeded internal function
e8ba2281f unix: remove support for FreeBSD < 10
78a1cd167 win,async: refactor internal implementation
7892bd6f7 win,tty: remove no longer needed workaround
c89ddacfa win: remove unused code
4613cc916 win: drop support for old MinGW versions
3118e5a4e unix,win: share loop watchers implementation
de0054b6d win: use queues for idle, prepare and check handles
3f21dd64d win: remove ABI compatibility hacks
9329327a2 win: drop support for Windows < Vista
9d2c77e9f win,build: drop support for Visual Studio < 2015
9066c89b9 macOS: don't dynamically load pthread_setname_np
46dbdd0f9 macOS: drop support for macOS < 10.9
d7f65293c core: add reserved fields to uv_loop_t
c82eedd0a unix: remove 1.x-compat code from fallback barrier
d85d606db Now working on version 2.0.0
3fd4e1c5e win: replace non-ansi characters in source file
49cb40c47 windows: remove duplicated field
```

**Comment by cjihrig** [comment]
OK. I guess I'll close this out. We can revisit in 9-10 months.

**Comment by santigimeno** [comment]
Sorry for reopening this. I'll give my 2 cents:

I'm not sure about the best way forward. I thing Ben's points to delay the release are valid but I think there are some things to also take into account:

- `nodejs` can be our main downstream project but it's not the only one. AFAIK, at least `julialang` is keeping its own fork from our `master` branch so they would benefit from a `v2` release. Also probably they'd be more willing to include their own patches upstream.

- As Ben mentioned we're very close to the `nodejs@10` release. That can be problematic specially if we can't make it and have to maintain 2 release lines. But wouldn't this be a problem anyway even if there was no near `nodejs` release? There are other downstream projects that would be affected by this release and we should give them some kind of support for the `v1.x` branch for a while (maybe a bugfixes-only policy)

- Not being this a big release can also make easier the maintenance of both branches.

- Contributors trying to add `major` changes to the project can be discouraged for a new large delay in the release.

**Comment by davisjam** [comment]
I'm just a bit player here but I agree with @santigimeno, especially on this: `Contributors trying to add major changes to the project can be discouraged for a new large delay in the release.`

**Comment by bnoordhuis** [comment]
> wouldn't this be a problem anyway even if there was no near nodejs release?

Not of the same magnitude because we could have dropped support end of 2019 when Node.js v8.x goes out of support.  Ample time for downstream projects to upgrade and still a far cry from the 3 years we're looking at now.

I don't disagree with your points but for me the prolonged maintenance angle is the most important.

**Comment by jasnell** [comment]
Does it have to wait for 12? Node.js 11.0.0 would give us a solid 6 month window.

**Comment by cjihrig** [comment]
I think the issue is more about Node 10 shipping with libuv 1.x and, as a consequence, us having to maintain libuv 1.x until April 2021 along with v2.x. Personally, I don't think the 1.x maintenance would be bad if we took a "security fixes and critical bugs only" approach, but I also don't want to speak for all of the other maintainers. That approach would also mean that Node 10 wouldn't get any libuv enhancements though.

**Comment by pfreixes** [comment]
So, once lost the Node.js 10 version train the question here is, should libuv wait 6/11 months for the new release?

Node.js is not the unique downstream,  others such us uvloop might be inreresteded in having the chance to use a new major release. Indeed, having a new major release not aligned with a  LTS Node.js release can be in some way an advantage, having a window time to make the proper battle testing.

Said that, IMHO wait for 9-10 months for the next call might be discouraging. Is there any chance to have a  plan for the following months?

**Comment by jebej** [comment]
The Julia project is also already using a v2 fork, and [would benefit](https://github.com/JuliaLang/julia/pull/25062#issuecomment-361140153) from a proper release.

**Comment by cjihrig** [comment]
How about this... there are currently [14 open PRs targeting v2](https://github.com/libuv/libuv/pulls?q=is%3Apr+is%3Aopen+label%3Av2). If you really want to see a v2 release (and I would really like to), help triage and review them. That way, we can figure out what will go into v2, and what can be closed or postponed.

**Comment by cjihrig** [comment]
I'd like to revisit this. By landing, closing, and excluding things, I've gotten the v2 target down to 4 PRs.

Node 11 is ~4 months away - with a change cutoff more like 3 months I imagine. We need to work through the remaining PRs, merge v1.x into master, and update Node. At that point we could put v1.x into maintenance mode.

There is also the matter of other consumers, like Julia, that would like to have a v2 release.

**Comment by bnoordhuis** [comment]
The basic issue hasn't changed though: we can't sunset v1.x for three more years and even moving it to maintenance mode isn't really an option for the next 22 months.

I won't block a v2.x release but I have to be parsimonious with my time and don't expect to be able to support both branches equally well.
